torch/fx/quantization.py

```diff
@@ -0,0 +1,329 @@
+            return False
+    elif node.target != self_match:
+        return False
+
+    if not arg_matches:
+        return True
+
+    if len(arg_matches) != len(node.args):
+        return False
+
+    return all(matches(modules, node, arg_match, max_uses=1) for node, arg_match in zip(node.args, arg_matches))
+
+
+class Quantizer:
+    def __init__(self, mod, patterns=DEFAULT_QUANTIZATION_PATTERNS, quant_ctor=DefaultQuant):
+        self.root = mod.root
+        self.graph = mod.graph
+        self.quant_ctor = quant_ctor
+
+        # cached information for observe
+        self.state_dict = self.root.state_dict()
+        self.modules = dict(self.root.named_modules())
+
+        # match the patterns that will get quantized
+        self.matches = self._find_matches(patterns)
+        # find _inputs_ to matched nodes that are not quantized, these
+        # have to be quantized, which requires measuring stats,
+        # initialize an quant_ctor object for each
+        self.quants = self._find_quants(quant_ctor)
+
+
+
+    def observe(self, args):
+        # most of this function is just an interpreter for the graph
+        # it would be possible to put this in some abstraction, but
+        # it is pretty nice to just be able to see exactly what is happening here
+        # and hack on it.
+        # maybe we should just provide an example interpreter that people copy/paste
+        # then edit.
+        args_iter = iter(args)
+        env = {}
+
+        def load_arg(a):
+            return map_arg(a, lambda node: env[node.name])
+
+        for node in self.graph.nodes:
+            if node.op == 'placeholder':
+                result = next(args_iter)
+            elif node.op == 'get_param':
+                result = self.state_dict[node.target]
+            elif node.op == 'call_function':
+                result = node.target(*load_arg(node.args), **load_arg(node.kwargs))
+            elif node.op == 'call_method':
+                self_obj, *args = load_arg(node.args)
+                kwargs = load_arg(node.kwargs)
+                result = getattr(self_obj, node.target)(*args, **kwargs)
+            elif node.op == 'call_module':
+                result = self.modules[node.target](*load_arg(node.args), **load_arg(node.kwargs))
+
+            env[node.name] = result
+            root_node, obj = self.matches.get(node.name, (None, None))
+            if root_node is node:
+                obj.observe(node, env)
+            if node.name in self.quants:
+                self.quants[node.name].observe(node, env)
+
+        return load_arg(self.graph.result)
+
+    def quantize(self):
+        self.quantized_graph = Graph()
+        self.delegate = DelegateBase(self.quantized_graph)
+
+        env = {}
+        quant_env = {}
+
+        def load_arg(n, quantized):
+            if not quantized:
+                if n.name not in env and n.name in quant_env:
+                    env[n.name] = quant_env[n.name].dequantize()
```

Comment:
This will be the only place where you need to construct a proxy `Proxy(quant_env[n.name]).dequantize().node`